elastic/elastic-agent
# Add `mage integration:test` target

## What does this PR do?



Adds the new `mage integration:test` target that uses the `testing/define` module and `testing/runner` module to perform integration and end-to-end testing of the Elastic Agent on remote hosts.

Updates the `mage integration:auth` target to create a service account, assign roles, and generate the service token key for authentication to GCP.

Calling `mage integration:test` works as follows:

1. Parses the test information using the `testing/define` module.
2. Passes the test information to the new `testing/runner.Runner`.
3. `Runner` parses the information and determines the instance its requires to perform the tests.
4. Generates SSH keys
5. Creates a zip archive of the current repo (including uncommitted changes)
6. Pulls the latest OGC image
7. Creates any Cloud instances required for the test (could be multiple)
8. Imports the instance layouts into OGC
9. Calls OGC up to deploy the instances
10. Calls OGC ls to gather IP address information
11. SSH into each instance
12. Runs apt-get update
13. Runs apt-get install build-essential unzip
14. Installs golang using the version defined in `.go-version`
15. Copies and extracts the repo zip created in step 5
16. Installs mage and gotestsum
17. Copies only the required package build of elastic-agent to the host
18. Runs the set of non-sudo tests capturing the results
19. Runs the set of sudo tests capturing the results
20. Merges the results into a unified output, JSON, and xml output

**NOTE**: At the moment it only supports Ubuntu, but does run on both amd64 and arm64.

## Why is it important?



Allows integration and end-to-end testing to be performed from a CI work or a local developers laptop.

## Checklist





## How to test this PR locally



First ensure that your authenticated with: `mage integration:auth`
Then run the tests: `mage integration:test`

## Logs



```
>>> Pulling latest ogc image
>>> Creating zip archive of repo to send to remote hosts
>>> Create SSH keys to use for SSH
>>> Creating ESS cloud 8.9.0-SNAPSHOT (at-blake-rouse-agent-testing-890-SNAPSHOT)
>>> Import layouts into ogc
>>> Bring up instances through ogc
>>> Starting SSH connection to instance ogc-chi3jdr511inftcgrn4g-286a @ 34.72.132.49
>>> Starting SSH connection to instance ogc-chi3jdr511inftcgrn1g-0952 @ 35.225.179.217
>>> Starting SSH connection to instance ogc-chi3jdr511inftcgrn40-a301 @ 34.30.89.237
>>> Starting SSH connection to instance ogc-chi3jdr511inftcgrn20-dc1a @ 34.170.81.58
>>> Connected over SSH to instance ogc-chi3jdr511inftcgrn4g-286a
>>> Preparing instance ogc-chi3jdr511inftcgrn4g-286a
>>> Running apt-get update on ogc-chi3jdr511inftcgrn4g-286a
>>> Connected over SSH to instance ogc-chi3jdr511inftcgrn1g-0952
>>> Preparing instance ogc-chi3jdr511inftcgrn1g-0952
>>> Running apt-get update on ogc-chi3jdr511inftcgrn1g-0952
>>> Connected over SSH to instance ogc-chi3jdr511inftcgrn20-dc1a
>>> Preparing instance ogc-chi3jdr511inftcgrn20-dc1a
>>> Running apt-get update on ogc-chi3jdr511inftcgrn20-dc1a
>>> Connected over SSH to instance ogc-chi3jdr511inftcgrn40-a301
>>> Preparing instance ogc-chi3jdr511inftcgrn40-a301
>>> Running apt-get update on ogc-chi3jdr511inftcgrn40-a301
>>> Install build-essential and unzip on ogc-chi3jdr511inftcgrn20-dc1a
>>> Install build-essential and unzip on ogc-chi3jdr511inftcgrn1g-0952
>>> Install build-essential and unzip on ogc-chi3jdr511inftcgrn40-a301
>>> Install golang 1.19.9 (arm64) on ogc-chi3jdr511inftcgrn20-dc1a
>>> Copying repo to ogc-chi3jdr511inftcgrn20-dc1a
>>> Running make mage and prepareOnRemote on ogc-chi3jdr511inftcgrn20-dc1a
>>> Install build-essential and unzip on ogc-chi3jdr511inftcgrn4g-286a
>>> Failed to install build-essential and unzip on ogc-chi3jdr511inftcgrn1g-0952; will wait 15 seconds and try again
>>> Failed to install build-essential and unzip on ogc-chi3jdr511inftcgrn40-a301; will wait 15 seconds and try again
>>> Copying agent build elastic-agent-8.9.0-linux-arm64.tar.gz to ogc-chi3jdr511inftcgrn20-dc1a
>>> Running apt-get update on ogc-chi3jdr511inftcgrn1g-0952
>>> Running apt-get update on ogc-chi3jdr511inftcgrn40-a301
>>> Install build-essential and unzip on ogc-chi3jdr511inftcgrn1g-0952
>>> Install build-essential and unzip on ogc-chi3jdr511inftcgrn40-a301
>>> Install golang 1.19.9 (amd64) on ogc-chi3jdr511inftcgrn1g-0952
>>> Install golang 1.19.9 (amd64) on ogc-chi3jdr511inftcgrn40-a301
>>> Install golang 1.19.9 (arm64) on ogc-chi3jdr511inftcgrn4g-286a
>>> Copying repo to ogc-chi3jdr511inftcgrn4g-286a
>>> Running make mage and prepareOnRemote on ogc-chi3jdr511inftcgrn4g-286a
>>> Copying repo to ogc-chi3jdr511inftcgrn40-a301
>>> Copying repo to ogc-chi3jdr511inftcgrn1g-0952
>>> Running make mage and prepareOnRemote on ogc-chi3jdr511inftcgrn1g-0952
>>> Running make mage and prepareOnRemote on ogc-chi3jdr511inftcgrn40-a301
>>> Copying agent build elastic-agent-8.9.0-linux-arm64.tar.gz to ogc-chi3jdr511inftcgrn4g-286a
>>> Instance ogc-chi3jdr511inftcgrn20-dc1a waiting for stack to be ready
>>> Instance ogc-chi3jdr511inftcgrn20-dc1a will continue stack is ready
>>> Starting tests on ogc-chi3jdr511inftcgrn20-dc1a
>>> Test output ogc-chi3jdr511inftcgrn20-dc1a (stderr): go: downloading github.com/rs/zerolog v1.27.0
>>> Test output on ogc-chi3jdr511inftcgrn20-dc1a (stdout): >> go test: remote-linux-arm64-ubuntu-2204.integration Testing
>>> Test output on ogc-chi3jdr511inftcgrn20-dc1a (stdout): exec: gotestsum --no-color -f standard-quiet --junitfile build/TEST-go-remote-linux-arm64-ubuntu-2204.integration.xml --jsonfile build/TEST-go-remote-linux-arm64-ubuntu-2204.integration.out.json -- -tags integration -test.run TestDiagnosticsCommandIntegrationTestSuite|TestFakeComponentIntegrationTestSuite github.com/elastic/elastic-agent/testing/integration
.... continues with out...
>>> Bring down instances through ogc
>>> Testing completed
>>> Console output written here: build/TEST-go-integration.out
>>> Console JSON output written here: build/TEST-go-integration.out.json
>>> JUnit XML written here: build/TEST-go-integration.xml
```
